readium/streamer/src/main/java/org/readium/r2/streamer/PublicationFactory.kt

```diff
@@ -0,0 +1,267 @@
+        onCreatePublication: Publication.Builder.() -> Unit = {},
+        warnings: WarningLogger? = null
+    ): Try<Publication, Publication.OpeningException> {
+        val builder = parse(publicationAsset, warnings)
+            .getOrElse { return Try.failure(wrapParserException(it)) }
+
+        builder.apply(onCreatePublication)
+
+        val publication = builder.build()
+            .apply { addLegacyProperties(publicationAsset.mediaType) }
+        return Try.success(publication)
+    }
+
+    private suspend fun parse(
+        publicationAsset: PublicationParser.Asset,
+        warnings: WarningLogger?
+    ): Try<Publication.Builder, PublicationParser.Error> {
+        for (parser in parsers) {
+            val result = parser.parse(publicationAsset, warnings)
+            if (
+                result is Try.Success ||
+                result is Try.Failure && result.value !is PublicationParser.Error.FormatNotSupported
+            ) {
+                return result
+            }
+        }
+        return Try.failure(PublicationParser.Error.FormatNotSupported())
+    }
+
+    private fun wrapParserException(e: PublicationParser.Error): Publication.OpeningException =
+        when (e) {
+            is PublicationParser.Error.FormatNotSupported ->
+                Publication.OpeningException.UnsupportedAsset("Cannot find a parser for this asset")
+            is PublicationParser.Error.IO ->
+                when (e.resourceError) {
+                    is Resource.Exception.BadRequest, is Resource.Exception.Other ->
+                        Publication.OpeningException.Unexpected(e)
+                    is Resource.Exception.Forbidden ->
+                        Publication.OpeningException.Forbidden(e)
+                    is Resource.Exception.NotFound ->
+                        Publication.OpeningException.ParsingFailed(e)
+                    is Resource.Exception.OutOfMemory ->
+                        Publication.OpeningException.OutOfMemory(e)
+                    is Resource.Exception.Unavailable, is Resource.Exception.Offline ->
+                        Publication.OpeningException.Unavailable(e)
+                }
+            is PublicationParser.Error.OutOfMemory ->
+                Publication.OpeningException.OutOfMemory(e)
+            is PublicationParser.Error.ParsingFailed ->
+                Publication.OpeningException.ParsingFailed(e)
+        }
+
+    @Suppress("UNCHECKED_CAST")
+    private suspend fun <T, R> List<T>.lazyMapFirstNotNullOrNull(transform: suspend (T) -> R): R? {
+        for (it in this) {
+            return transform(it) ?: continue
+        }
+        return null
+    }
+
+    private fun Publication.addLegacyProperties(mediaType: MediaType?) {
+        @Suppress("DEPRECATION")
+        type = mediaType.toPublicationType()
+
+        if (mediaType == MediaType.EPUB)
+            setLayoutStyle()
+    }
+}
+
+internal fun MediaType?.toPublicationType(): Publication.TYPE =
```

Comment:
I guess we don't need this anymore, as `Publication.TYPE` is deprecated with error level?

Reply:
In the next PR only!